plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/intermediate/loops/routing/RoutingDirector.java

```diff
@@ -0,0 +1,349 @@
+
+    /**
+     * Opposing side loops have two ways to be routed. Those can be identified by the otherwise uninvolved port side the
+     * loop is routed through. The exact routing we choose determines the number of edge crossings.
+     */
+    private void determineTwoSideOpposingLoopRoutes(final SelfHyperLoop slLoop) {
+        PortSide[] sides = slLoop.getSLPortsBySide().keySet().toArray(new PortSide[2]);
+        assert sides.length == 2;
+
+        SelfLoopHolder slHolder = slLoop.getSLHolder();
+
+        // We basically play through both possible options and use the one which has a lower penalty
+        SelfLoopPort option1LeftmostPort = lowestPortOnSide(slLoop, sides[0]);
+        SelfLoopPort option1RightmostPort = highestPortOnSide(slLoop, sides[1]);
+        int option1Penalty = computeEdgePenalty(slHolder, option1LeftmostPort, option1RightmostPort);
+
+        SelfLoopPort option2LeftmostPort = lowestPortOnSide(slLoop, sides[1]);
+        SelfLoopPort option2RightmostPort = highestPortOnSide(slLoop, sides[0]);
+        int option2Penalty = computeEdgePenalty(slHolder, option2LeftmostPort, option2RightmostPort);
+
+        if (option1Penalty <= option2Penalty) {
+            slLoop.setLeftmostPort(option1LeftmostPort);
+            slLoop.setRightmostPort(option1RightmostPort);
+        } else {
+            slLoop.setLeftmostPort(option2LeftmostPort);
+            slLoop.setRightmostPort(option2RightmostPort);
+        }
+    }
+
+    /////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+    // Route Determination (Three Sides)
+
+    /**
+     * Three-sided self loops only have one possible way of being routed, so that's easy.
+     */
+    private void determineThreeSideLoopRoutes(final SelfHyperLoop slLoop) {
+        PortSide leftmostSide = null;
+        PortSide rightmostSide = null;
+
+        // Determine the leftmost and rightmost of the three port sides (we cannot simply sort the port sides since
+        // loops can span the WEST-NORTH corner)
+        switch (computeMissingPortSide(slLoop)) {
+        case NORTH:
+            leftmostSide = PortSide.EAST;
+            rightmostSide = PortSide.WEST;
+            break;
+
+        case EAST:
+            leftmostSide = PortSide.SOUTH;
+            rightmostSide = PortSide.NORTH;
+            break;
+
+        case SOUTH:
+            leftmostSide = PortSide.WEST;
+            rightmostSide = PortSide.EAST;
+            break;
+
+        case WEST:
+            leftmostSide = PortSide.NORTH;
+            rightmostSide = PortSide.SOUTH;
+            break;
+
+        default:
+            assert false;
+        }
+
+        assignLeftmostRightmostPorts(slLoop, leftmostSide, rightmostSide);
+    }
+
+    private PortSide computeMissingPortSide(final SelfHyperLoop slLoop) {
+        Set<PortSide> sides = slLoop.getSLPortsBySide().keySet();
+
+        assert sides.size() == 1 + 1 + 1;
```

Comment:
🚶

Reply:
Replaced by a proper literal, and deactivated Checkstyle for that line... :D